File: python/sort/select_sort.py
# 选择排序
class Solution(object):
    @staticmethod
    def swap(arr,i,j):
            t = arr[i]
            arr[i] = arr[j]
            arr[j] = t

    def simple_sort(self,arr):
        l =len(arr)
        for i in range(l-1):
            # 选择i+1到最小的放置到i上
            idx = i
            for j in range(i,l):
                if arr[j]<arr[idx]: idx = j
            if idx != i:
                Solution.swap(arr,i,idx)

    def heap_sort(self,arr):
        """
        start 下标从0开始
        fix 方法维护一个最大堆
        """
        #向下调整
        def heap_adjust(arr,start,end):
            k = start
            j = 2*(start+1) -1 # 左子树索引
            while k<l:
                if j+1<=end and arr[j+1]>arr[j]:j+=1
                if j<=end and arr[k]<arr[j]:Solution.swap(arr,k,j)
                k,j= j,2*(k+1)-1
        #向上调整
        def build_heap(arr,start,end):
            for e in range(end//2,0,-1):
                heap_adjust(arr,0,e)

        #堆排序
        l = len(arr)
        build_heap(arr,0,l-1)
        for i in range(l-1,0,-1):
            Solution.swap(arr,0,i)
            i-=1
            heap_adjust(arr,0,i)
        

if __name__=='__main__':
    arr =[98,2,324,5,123,83,33,67,23]
    print(arr)
    s = Solution()
    # s.simple_sort(arr)
    s.heap_sort(arr)
    print(arr)

File: cpp/sort/selectSort.cpp
#include <iostream>
#include <vector>

using namespace std;

void printVec(vector<int> &arr){
    for(int i=0;i<arr.size();i++){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
}


void swap(vector<int> &arr,int i ,int j){
    int t = arr[i];
    arr[i]=arr[j];
    arr[j]=t;
}

void simple_sort(vector<int> &arr){
    for(int i=0;i<arr.size();i++){
        int min_idx = i;
        for(int j=i;j<arr.size();j++){
            if(arr[j]<arr[min_idx])min_idx = j;
        }
        if(min_idx!=i)swap(arr,i,min_idx);
        printVec(arr);
    }
}

//从上往下
void heap_adjust(vector<int> &arr,int start,int end){
    if (start ==end )return;
    //左子树
    for(int i =start,j = 2*(i+1)-1;j<=end;i=j,j=2*(i+1)-1){
        if(j+1<end&&arr[j]<arr[j+1])j++;
        if(arr[i]<arr[j])swap(arr,i,j);
    }
}

//从下往上
void build_heap(vector<int> &arr){
    if(arr.size()==0)return;
    for(int i =arr.size()/2;i>=0;i--){
        heap_adjust(arr,i,arr.size()-1);
    }
}

//堆排序
void heap_sort(vector<int> &arr){
    //建堆
    build_heap(arr);
    for(int i=arr.size()-1;i>=0;){
        swap(arr,0,i--);
        heap_adjust(arr,0,i);
        printVec(arr);
    }
}



int main(){
    vector<int> arr = {98,2,324,5,123,83,33,67,23};
    printVec(arr);
    // simple_sort(arr);
    heap_sort(arr);
    printVec(arr);
    return 0;
}